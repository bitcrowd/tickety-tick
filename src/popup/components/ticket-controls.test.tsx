import { fireEvent, render } from "@testing-library/react";
import React from "react";

import { ticket as make } from "../../../test/factories";
import type { Props } from "./ticket-controls";
import TicketControls from "./ticket-controls";

describe("ticket-controls", () => {
  function subject(overrides: Partial<Props>) {
    const defaults: Props = {
      tickets: ["raz", "dva", "tri"].map((title, i) =>
        make({ id: `${i + 1}`, title }),
      ),
    };
    const props = { ...defaults, ...overrides };
    return render(<TicketControls {...props} />);
  }

  it("renders for multiple tickets", () => {
    expect(subject({}).asFragment()).toMatchSnapshot();
  });

  it("renders for a single ticket", () => {
    const tickets = [make({ id: "1", title: "raz" })];
    expect(subject({ tickets })).toMatchSnapshot();
  });

  it("renders a dropdown if there is more than one ticket", () => {
    const tickets = ["raz", "dva", "tri"].map((title, i) =>
      make({ id: `${i + 1}`, title }),
    );
    const screen = subject({ tickets });
    const select = screen.getByRole("combobox");
    expect(select).not.toBeDisabled();
  });

  it("renders a disabled dropdown if there is only one ticket", () => {
    const tickets = [make({ id: "2", title: "dva" })];
    const screen = subject({ tickets });
    const select = screen.getByRole("combobox");
    expect(select).toBeDisabled();
  });

  it("renders copy buttons for branch, commit and command of the selected ticket", () => {
    const ticket1 = make({ id: "1", title: "raz" });
    const ticket2 = make({ id: "2", title: "dva" });
    const screen = subject({ tickets: [ticket1, ticket2] });
    const select = screen.getByRole("combobox");
    let buttons = screen.getAllByRole("button") as HTMLButtonElement[];

    expect(buttons).toHaveLength(3);
    expect(buttons.map((button) => button.value)).toEqual([
      ticket1.fmt.branch,
      ticket1.fmt.commit,
      ticket1.fmt.command,
    ]);

    // select the second ticket (index = 1)
    fireEvent.change(select, { target: { value: 1 } });

    buttons = screen.getAllByRole("button") as HTMLButtonElement[];
    expect(buttons).toHaveLength(3);
    expect(buttons.map((button) => button.value)).toEqual([
      ticket2.fmt.branch,
      ticket2.fmt.commit,
      ticket2.fmt.command,
    ]);
  });

  it("allows to cycle through the buttons and the select via the tab key", () => {
    const screen = subject({});
    const buttons = screen.getAllByRole("button");
    const select = screen.getByRole("combobox");

    buttons.forEach((button) => {
      expect(button).toHaveAttribute("tabIndex", "1");
    });
    expect(select).toHaveAttribute("tabIndex", "1");
  });
});
