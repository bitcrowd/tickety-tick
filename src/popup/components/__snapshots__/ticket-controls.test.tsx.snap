// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ticket-controls renders for a single ticket 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="d-flex flex-row mb-2 btn-group btn-group-lg"
          role="group"
        >
          <button
            class="d-flex justify-content-center align-items-center btn btn-primary"
            tabindex="1"
            title="Copy branch name"
            type="button"
            value="branch-1"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-git-branch"
              fill="currentColor"
              focusable="false"
              height="18"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="18"
            >
              <path
                d="M9.5 3.25a2.25 2.25 0 1 1 3 2.122V6A2.5 2.5 0 0 1 10 8.5H6a1 1 0 0 0-1 1v1.128a2.251 2.251 0 1 1-1.5 0V5.372a2.25 2.25 0 1 1 1.5 0v1.836A2.493 2.493 0 0 1 6 7h4a1 1 0 0 0 1-1v-.628A2.25 2.25 0 0 1 9.5 3.25Zm-6 0a.75.75 0 1 0 1.5 0 .75.75 0 0 0-1.5 0Zm8.25-.75a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM4.25 12a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5Z"
              />
            </svg>
            <span
              class="ps-1 small btn-label btn-label-conceal"
            >
              Branch
            </span>
          </button>
          <button
            class="d-flex justify-content-center align-items-center btn btn-primary"
            tabindex="1"
            title="Copy commit message"
            type="button"
            value="commit-1"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-comment"
              fill="currentColor"
              focusable="false"
              height="18"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="18"
            >
              <path
                d="M1 2.75C1 1.784 1.784 1 2.75 1h10.5c.966 0 1.75.784 1.75 1.75v7.5A1.75 1.75 0 0 1 13.25 12H9.06l-2.573 2.573A1.458 1.458 0 0 1 4 13.543V12H2.75A1.75 1.75 0 0 1 1 10.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h4.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z"
              />
            </svg>
            <span
              class="ps-1 small btn-label btn-label-conceal"
            >
              Message
            </span>
          </button>
          <button
            class="d-flex justify-content-center align-items-center btn btn-primary"
            tabindex="1"
            title="Copy CLI command"
            type="button"
            value="command-1"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-terminal"
              fill="currentColor"
              focusable="false"
              height="18"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="18"
            >
              <path
                d="M0 2.75C0 1.784.784 1 1.75 1h12.5c.966 0 1.75.784 1.75 1.75v10.5A1.75 1.75 0 0 1 14.25 15H1.75A1.75 1.75 0 0 1 0 13.25Zm1.75-.25a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h12.5a.25.25 0 0 0 .25-.25V2.75a.25.25 0 0 0-.25-.25ZM7.25 8a.749.749 0 0 1-.22.53l-2.25 2.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L5.44 8 3.72 6.28a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l2.25 2.25c.141.14.22.331.22.53Zm1.5 1.5h3a.75.75 0 0 1 0 1.5h-3a.75.75 0 0 1 0-1.5Z"
              />
            </svg>
            <span
              class="ps-1 small btn-label btn-label-conceal"
            >
              Command
            </span>
          </button>
        </div>
        <div>
          <select
            class="form-control form-control-sm"
            disabled=""
            tabindex="1"
          >
            <option
              value="0"
            >
              raz
            </option>
          </select>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="d-flex flex-row mb-2 btn-group btn-group-lg"
        role="group"
      >
        <button
          class="d-flex justify-content-center align-items-center btn btn-primary"
          tabindex="1"
          title="Copy branch name"
          type="button"
          value="branch-1"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-git-branch"
            fill="currentColor"
            focusable="false"
            height="18"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="18"
          >
            <path
              d="M9.5 3.25a2.25 2.25 0 1 1 3 2.122V6A2.5 2.5 0 0 1 10 8.5H6a1 1 0 0 0-1 1v1.128a2.251 2.251 0 1 1-1.5 0V5.372a2.25 2.25 0 1 1 1.5 0v1.836A2.493 2.493 0 0 1 6 7h4a1 1 0 0 0 1-1v-.628A2.25 2.25 0 0 1 9.5 3.25Zm-6 0a.75.75 0 1 0 1.5 0 .75.75 0 0 0-1.5 0Zm8.25-.75a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM4.25 12a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5Z"
            />
          </svg>
          <span
            class="ps-1 small btn-label btn-label-conceal"
          >
            Branch
          </span>
        </button>
        <button
          class="d-flex justify-content-center align-items-center btn btn-primary"
          tabindex="1"
          title="Copy commit message"
          type="button"
          value="commit-1"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-comment"
            fill="currentColor"
            focusable="false"
            height="18"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="18"
          >
            <path
              d="M1 2.75C1 1.784 1.784 1 2.75 1h10.5c.966 0 1.75.784 1.75 1.75v7.5A1.75 1.75 0 0 1 13.25 12H9.06l-2.573 2.573A1.458 1.458 0 0 1 4 13.543V12H2.75A1.75 1.75 0 0 1 1 10.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h4.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z"
            />
          </svg>
          <span
            class="ps-1 small btn-label btn-label-conceal"
          >
            Message
          </span>
        </button>
        <button
          class="d-flex justify-content-center align-items-center btn btn-primary"
          tabindex="1"
          title="Copy CLI command"
          type="button"
          value="command-1"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-terminal"
            fill="currentColor"
            focusable="false"
            height="18"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="18"
          >
            <path
              d="M0 2.75C0 1.784.784 1 1.75 1h12.5c.966 0 1.75.784 1.75 1.75v10.5A1.75 1.75 0 0 1 14.25 15H1.75A1.75 1.75 0 0 1 0 13.25Zm1.75-.25a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h12.5a.25.25 0 0 0 .25-.25V2.75a.25.25 0 0 0-.25-.25ZM7.25 8a.749.749 0 0 1-.22.53l-2.25 2.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L5.44 8 3.72 6.28a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l2.25 2.25c.141.14.22.331.22.53Zm1.5 1.5h3a.75.75 0 0 1 0 1.5h-3a.75.75 0 0 1 0-1.5Z"
            />
          </svg>
          <span
            class="ps-1 small btn-label btn-label-conceal"
          >
            Command
          </span>
        </button>
      </div>
      <div>
        <select
          class="form-control form-control-sm"
          disabled=""
          tabindex="1"
        >
          <option
            value="0"
          >
            raz
          </option>
        </select>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ticket-controls renders for multiple tickets 1`] = `
<DocumentFragment>
  <form>
    <div
      class="d-flex flex-row mb-2 btn-group btn-group-lg"
      role="group"
    >
      <button
        class="d-flex justify-content-center align-items-center btn btn-primary"
        tabindex="1"
        title="Copy branch name"
        type="button"
        value="branch-1"
      >
        <svg
          aria-hidden="true"
          class="octicon octicon-git-branch"
          fill="currentColor"
          focusable="false"
          height="18"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="18"
        >
          <path
            d="M9.5 3.25a2.25 2.25 0 1 1 3 2.122V6A2.5 2.5 0 0 1 10 8.5H6a1 1 0 0 0-1 1v1.128a2.251 2.251 0 1 1-1.5 0V5.372a2.25 2.25 0 1 1 1.5 0v1.836A2.493 2.493 0 0 1 6 7h4a1 1 0 0 0 1-1v-.628A2.25 2.25 0 0 1 9.5 3.25Zm-6 0a.75.75 0 1 0 1.5 0 .75.75 0 0 0-1.5 0Zm8.25-.75a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM4.25 12a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5Z"
          />
        </svg>
        <span
          class="ps-1 small btn-label btn-label-conceal"
        >
          Branch
        </span>
      </button>
      <button
        class="d-flex justify-content-center align-items-center btn btn-primary"
        tabindex="1"
        title="Copy commit message"
        type="button"
        value="commit-1"
      >
        <svg
          aria-hidden="true"
          class="octicon octicon-comment"
          fill="currentColor"
          focusable="false"
          height="18"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="18"
        >
          <path
            d="M1 2.75C1 1.784 1.784 1 2.75 1h10.5c.966 0 1.75.784 1.75 1.75v7.5A1.75 1.75 0 0 1 13.25 12H9.06l-2.573 2.573A1.458 1.458 0 0 1 4 13.543V12H2.75A1.75 1.75 0 0 1 1 10.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h4.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z"
          />
        </svg>
        <span
          class="ps-1 small btn-label btn-label-conceal"
        >
          Message
        </span>
      </button>
      <button
        class="d-flex justify-content-center align-items-center btn btn-primary"
        tabindex="1"
        title="Copy CLI command"
        type="button"
        value="command-1"
      >
        <svg
          aria-hidden="true"
          class="octicon octicon-terminal"
          fill="currentColor"
          focusable="false"
          height="18"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="18"
        >
          <path
            d="M0 2.75C0 1.784.784 1 1.75 1h12.5c.966 0 1.75.784 1.75 1.75v10.5A1.75 1.75 0 0 1 14.25 15H1.75A1.75 1.75 0 0 1 0 13.25Zm1.75-.25a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h12.5a.25.25 0 0 0 .25-.25V2.75a.25.25 0 0 0-.25-.25ZM7.25 8a.749.749 0 0 1-.22.53l-2.25 2.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L5.44 8 3.72 6.28a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l2.25 2.25c.141.14.22.331.22.53Zm1.5 1.5h3a.75.75 0 0 1 0 1.5h-3a.75.75 0 0 1 0-1.5Z"
          />
        </svg>
        <span
          class="ps-1 small btn-label btn-label-conceal"
        >
          Command
        </span>
      </button>
    </div>
    <div>
      <select
        class="form-control form-control-sm"
        tabindex="1"
      >
        <option
          value="0"
        >
          raz
        </option>
        <option
          value="1"
        >
          dva
        </option>
        <option
          value="2"
        >
          tri
        </option>
      </select>
    </div>
  </form>
</DocumentFragment>
`;
